### PR TITLE
Common contacts

### PR DESCRIPTION
**Pull request by msultan**
- [x] Implement feature / fix bug
 - [x] Add tests
 - [x] Update changelog

New featurizer that can handle multiple protein sequences. The transform object internally
figures out the correct residue indices needed for each protein sequence in the alignment. 

```python
feat = CommonContactFeaturizer(alignment_file="./alignment.fst",\
wanted_positions=[0,12,1,14,123,121,231])

t1 = md.load("protein1.pdb")
t2 = md.load("protein2.pdb")

f1 = feat.transform(t1)
f2 = feat.transform(t2)
```

### CODE REVIEW

**Comment by mpharrigan** [comment]
What needs to be done? Why WIP?

**Comment by msultan** [comment]
@mpharrigan This version needs tests + docs + code cleanup. Also, I currently have a ```wanted_positions``` kwarg which is used to define _all_ pairwise contact features. However, the regular ```ContactFeaturizer``` takes in pairs of residue indices instead of residue indices.

File: msmbuilder/featurizer/featurizer.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 import itertools
 from sklearn.base import TransformerMixin
 from sklearn.externals.joblib import Parallel, delayed
-
+from Bio import SeqIO
```
Comment (cxhernandez) [inline msmbuilder/featurizer/featurizer.py]: Have a try-except for users that don't have `biopython`. It's not a requirement for MSMBuilder, right?

**Comment by cxhernandez** [comment]
What do you think about it accepting a list of aligned sequences instead of a filename?

**Comment by msultan** [comment]
> What do you think about it accepting a list of aligned sequences instead of a filename?
good call, it will get rid of the biopython requirements too.

**Comment by msultan** [comment]
So this doesnt work with sequences that have capped termini primarily because mdtraj's ```to_fasta``` method gives that amino acid Nonetype. Right now, i am raising exceptions all over the place if the sequences dont match up to the alignment but i think there should be a more elegant solution to this.

**Comment by msultan** [comment]
this is ready for code review.

**Comment by msultan** [comment]
changed logic to not subsample contacts. If fewer contacts are required, the```wanted_locs``` can be subsampled. swapped out one of the tests for a randomized version.

**Comment by msultan** [comment]
@mpharrigan @cxhernandez @brookehus anyone wanna take a stab at checking out the code?

File: docs/changelog.rst [hunk 2]
```diff
@@ -26,6 +26,10 @@ New Features
 - ``BinaryContactFeaturizer`` featurizes a trajectory into a
   boolean array corresponding to whether each residue-residue
   distance is below a cutoff (#798).
+- ``CommonContactFeaturizer`` featurizes a bunch of different
+  trajectories,with different topologies, using a common set of
+  inter-residue contacts. Only featurizers on aminoacids >3
+  positions apart in sequence and requires an alignment to work(#876).
```
Comment (brookehus) [inline docs/changelog.rst]: Does this need to be changed since you aren't subsampling by 3 anymore?

File: msmbuilder/featurizer/featurizer.py [hunk 3]
```diff
@@ -1000,6 +1001,182 @@ def describe_features(self, traj):
         return feature_descs
 
 
+class CommonContactFeaturizer(Featurizer):
+    """Featurizer based on residue-residue contacts between
+    desired positions of an alignment file.
```
Comment (brookehus) [inline msmbuilder/featurizer/featurizer.py]: I think this is supposed to be one line e.g. for `msmb -h`

File: msmbuilder/featurizer/featurizer.py [hunk 4]
```diff
@@ -1000,6 +1001,182 @@ def describe_features(self, traj):
...
+            'ca' : distance between two residues is given by the distance
+                between their alpha carbons
+            'closest' : distance is the closest distance between any
+                two atoms in the residues
+            'closest-heavy' : distance is the closest distance between
+                any two non-hydrogen atoms in the residues
+    ignore_nonprotein : bool
+        When using `contact==all`, don't compute contacts between
+        "residues" which are not protein (i.e. do not contain an alpha
+        carbon).
```
Comment (brookehus) [inline msmbuilder/featurizer/featurizer.py]: Can you put in default parameters when relevant?

File: msmbuilder/featurizer/featurizer.py [hunk 5]
```diff
@@ -1000,6 +1001,182 @@ def describe_features(self, traj):
...
+"""
+
+    def __init__(self, alignment=None, wanted_positions='all',
+                 same_residue='True',
+                  scheme='closest-heavy', ignore_nonprotein=True):
```
Comment (brookehus) [inline msmbuilder/featurizer/featurizer.py]: Where is `conserved_only`?

File: msmbuilder/featurizer/featurizer.py [hunk 6]
```diff
@@ -1000,6 +1001,182 @@ def describe_features(self, traj):
...
+        elif type(alignment)==dict:
+            self.alignment = alignment
+        else:
+            raise ValueError("Alignment is not of type list or dict.")
+
+        self.protein_list = list(self.alignment.keys())
+        self.same_residue = same_residue
+        self.wanted_positions = wanted_positions
+        if wanted_positions is 'all':
+            #use the max length(probably a horrible idea)
```
Comment (brookehus) [inline msmbuilder/featurizer/featurizer.py]: Why?

File: msmbuilder/featurizer/featurizer.py [hunk 7]
```diff
@@ -1000,6 +1001,182 @@ def describe_features(self, traj):
...
+    def describe_features(self, traj):
+        seq_id = [k for k,v in self.all_sequences.items()
+                  if v == traj.top.to_fasta(chain=0)][0]
+        assert traj.top.chain(0).n_residues==len(self.all_sequences[seq_id])
+        return self.feat_dict[seq_id].describe_features(traj)
+
+    def _map_residue_ind_seq_ind(self,alignment):
+        """
+
+        Returns
```
Comment (brookehus) [inline msmbuilder/featurizer/featurizer.py]: update what it returns?

**Comment by brookehus** [comment]
I think I've expressed this to all of you in person but for the record, I think, in the style of #883, we should make a new thing in the `featurizer` directory, e.g. `multiseq.py`, which makes a base class for multi sequence featurizers and then contains all the featurizers that require multiple sequences.

**Comment by msultan** [comment]
@brookehus is this what you had in mind with the multiple files

**Comment by mpharrigan** [comment]
Merged in manually in a47622cb6d7f766d8fd253f82e48b419d0d5ae23

**Pull request closed without merging**